salt/utils/gitfs.py

```diff
@@ -3045,6 +3045,56 @@ def checkout(self):
                 else:
                     self.pillar_dirs[cachedir] = env
 
+    @contextlib.contextmanager
+    def global_lock(self, timeout=60, poll_interval=0.5):
+        '''
+        Set and automatically clear a lock on all the repos
+        '''
+
+        # Make sure that we have a positive integer timeout, otherwise just set
+        # it to zero.
+        try:
+            timeout = int(timeout)
+        except ValueError:
+            timeout = 0
+        else:
+            if timeout < 0:
+                timeout = 0
+
+        if not isinstance(poll_interval, (six.integer_types, float)) \
+                or poll_interval < 0:
+            poll_interval = 0.5
+
+        if poll_interval > timeout:
+            poll_interval = timeout
+
+        locked_repo = []
+        try:
+            time_start = time.time()
+            for repo in self.remotes:
+                while True:
+                    try:
+                        repo._lock(lock_type='global', failhard=True)
+                        locked_repo.append(repo)
+                        yield
+                        # Break out of his loop once we've yielded the lock, to
+                        # avoid continued attempts to iterate and establish lock
+                        break
+                    except (OSError, IOError, GitLockError) as exc:
+                        if not timeout or time.time() - time_start > timeout:
+                            raise GitLockError(exc.errno, exc.strerror)
+                        else:
+                            log.debug(
+                                'A global lock is already present for %s remote '
+                                '\'%s\', sleeping %f second(s)',
+                                self.role, remote.id, poll_interval
+                            )
+                            time.sleep(poll_interval)
+                            continue
```

Comment:
The yield needs to be outside of the for loop after you have locked all of the repos. Otherwise you yield multiple times once for each repo. The purpose of this function is to yield once after all repos are locked.
```suggestion
                        break
                    except (OSError, IOError, GitLockError) as exc:
                        if not timeout or time.time() - time_start > timeout:
                            raise GitLockError(exc.errno, exc.strerror)
                        else:
                            log.debug(
                                'A global lock is already present for %s remote '
                                '\'%s\', sleeping %f second(s)',
                                self.role, remote.id, poll_interval
                            )
                            time.sleep(poll_interval)
                            continue
            yield
```